### PR TITLE
Improve errors, add error codes

### PR DESCRIPTION
**Pull request by v0idpwn**
This changes a lot of errors from being atoms handled ad-hoc to be
structured errors. Structured errors allow us to have more
error-associated behavior, for example, every error can have a code, and
associated information to be used in logs. These structured errors also
allow us to have shared code paths in the caller code, reducing the
complexity of error handling in core parts of the codebase (e.g.: the
connection state machines).

This PR introduces an error behaviour (which is used alongside the
Exception behaviour) which has callbacks for log message and protocol
message (to be sent on the wire if this error happens for a client
connection), and an Error use-macro which makes the behaviour
implementation simple and does basic validation (does this error have a
code?).

It also introduces a new assertion, which we can use on returned errors,
to ensure that they are loggable and won't crash with the behaviour 
implementation. I haven't implemented it everywhere yet, but I think
it's a good addition for coverage.

Naturally, this PR touches a lot of surface.

## Summary by CodeRabbit

* **New Features**
  * Standardized, structured error types with machine-readable codes and a new API to expose the max startup packet size.

* **Bug Fixes**
  * More explicit, consistent error responses for auth, SSL/handshake, subscription, pool/worker not-found, checkout timeouts and client-limit conditions.
  * Improved PostgreSQL protocol error formatting and unified error-handling pathways.

* **Tests**
  * Expanded test coverage to assert new error shapes and startup/checkout edge cases; added validation helpers for error formatting.

### CODE REVIEW

File: lib/supavisor.ex [hunk 1]
```diff
@@ -62,32 +62,32 @@ defmodule Supavisor do
   def stop(id) do
     case get_global_sup(id) do
       nil ->
-        {:error, :tenant_not_found}
+        {:error, :tenant_not_running}
```
Comment (v0idpwn) [inline lib/supavisor.ex]: todo

File: lib/supavisor/client_handler.ex [hunk 2]
```diff
@@ -894,57 +800,30 @@ defmodule Supavisor.ClientHandler do
...
-
-        error_message = Server.encode_error_message(error_map)
-        HandlerHelpers.sock_send(data.sock, error_message)
-        exit(:normal)
-
-      {:error, {:exit, e}} ->
-        exit(e)
+      {:ok, {data.pool, db_pid, db_sock}}
     end
   end
```
[the remark this answers is not in the document]
Comment (v0idpwn) [reply lib/supavisor/client_handler.ex:816]: I'll remove this link, since the DbHandler already links on checkout.

File: lib/supavisor/errors/duplicate_prepared_statement_error.ex [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  @impl Supavisor.Error
+  def error_message(%{name: name}) do
+    "prepared statement #{inspect(name)} already exists"
+  end
+
+  @impl Supavisor.Error
+  def postgres_error(error) do
+    Supavisor.Error.protocol_error("FATAL", "42P05", message(error))
+  end
```
[the remark this answers is not in the document]
Comment (v0idpwn) [reply lib/supavisor/errors/duplicate_prepared_statement_error.ex:21]: For us it's fatal.

File: lib/supavisor/errors/prepared_statement_not_found_error.ex [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+  @impl Supavisor.Error
+  def error_message(%{name: name}) do
+    "prepared statement #{inspect(name)} does not exist"
+  end
+
+  @impl Supavisor.Error
+  def postgres_error(error) do
+    Supavisor.Error.protocol_error("FATAL", "26000", message(error))
+  end
```
[the remark this answers is not in the document]
Comment (v0idpwn) [reply lib/supavisor/errors/prepared_statement_not_found_error.ex:21]: Recovery is not something we want to deal with, if this happens we terminate the connection

File: test/support/asserts.ex [hunk 5]
```diff
@@ -42,4 +42,114 @@ defmodule Supavisor.Asserts do
...
+          {:error, exception} when is_exception(exception) ->
+            exception
+
+          exception when is_exception(exception) ->
+            exception
+
+          other ->
+            reraise ExUnit.AssertionError,
+                    "expected a exception or error tuple, got #{inspect(other)}"
+        end
```
[the remark this answers is not in the document]
Comment (v0idpwn) [reply test/support/asserts.ex:77]: This was fixed. But in the other places we should use reraise, just need to use reraise properly.

File: lib/supavisor/client_handler.ex [hunk 6]
```diff
@@ -1071,45 +951,43 @@ defmodule Supavisor.ClientHandler do
...
+    {:ok, addr} = HandlerHelpers.addr_from_sock(sock)
 
-        state == :handshake ->
-          :warning
+    if HandlerHelpers.filter_cidrs(info.tenant.allow_list, addr) == [] do
+      {:error, %AddressNotAllowedError{address: addr}}
+    else
+      :ok
+    end
+  end
```
[the remark this answers is not in the document]
Comment (v0idpwn) [reply lib/supavisor/client_handler.ex:965]: While this is a valid concern, we currently don't handle this sort of error in Supavisor. It should be reserved for future improvements.

[the remark this answers is not in the document]
Comment (v0idpwn) [reply lib/supavisor/client_handler.ex:965]: No.

File: lib/supavisor/errors/startup_packet_too_large_error.ex [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+  @type t() :: %__MODULE__{
+          packet_size: non_neg_integer(),
+          code: binary()
+        }
+
+  @impl Supavisor.Error
+  def error_message(%{packet_size: packet_size}) do
+    "Startup packet too large: #{packet_size} bytes (max 1024 bytes)"
+  end
```
[the remark this answers is not in the document]
Comment (v0idpwn) [reply lib/supavisor/errors/startup_packet_too_large_error.ex:16]: This was resolved in [d6420ab](https://github.com/supabase/supavisor/pull/841/commits/d6420ab5a339bf640c5af157458f386d374b68bd)

File: lib/supavisor/errors/max_connections_error.ex
Comment (mentels) [inline lib/supavisor/errors/max_connections_error.ex:1]: This one is difficult to reason about:
* for `:transaction` and `:proxy` connections we construct the log message manually
* for the `:session` we relay on the default `message/1` implementation that will also prepend the `:code` in front of the `error_message/1`

Why cannot we just relay on the `error_message/1` here? We have all the context we log, already there.

File: lib/supavisor/errors/max_pools_reached_error.ex
Comment (mentels) [inline lib/supavisor/errors/max_pools_reached_error.ex:1]: could be useful to add the `:limit`

File: lib/supavisor/client_handler/error.ex [hunk 8]
```diff
@@ -3,306 +3,79 @@ defmodule Supavisor.ClientHandler.Error do
...
+    if context == :handshake do
+      Telem.client_join(:fail, data.id)
+    end
 
-  defp process({:error, :max_pools_reached}, _context) do
-    %{
-      error: Server.error_message("XX000", "Max pools count reached"),
-      log_message: "Max pools count reached"
-    }
+    {:stop, :normal}
```
Comment (mentels) [inline lib/supavisor/client_handler/error.ex:50]: That's weird that we send `ReadyForQuery` and then stop the connection, why is that?

File: test/integration/proxy_test.exs [hunk 9]
```diff
@@ -237,14 +254,48 @@ defmodule Supavisor.Integration.ProxyTest do
...
+
+    connection_opts = [
+      hostname: db_conf[:hostname],
+      port: Application.get_env(:supavisor, :proxy_port_transaction),
+      username: "checkout_timeout_test.proxy_tenant1",
+      database: "postgres",
+      password: "postgres"
+    ]
+
+    # Start 2 connections and keep them busy with transactions (pool_size is 2)
```
Comment (mentels) [inline test/integration/proxy_test.exs:275]: In `supavisor/priv/repo/seeds_after_migration.exs` the `pool_size` is set to `9`

The comment says it's `2`, but only the 4th connection times out, what is. going on here? :D

File: lib/supavisor/client_handler/error.ex (hunk 8 above)
Comment (v0idpwn) [reply lib/supavisor/client_handler/error.ex:50]: It's a quirk of the postgres protocol. During authentication, we should send the error and close the connection, without a ready for query. But after auth:

Simple query protocol docs says:

>In the event of an error, ErrorResponse is issued followed by ReadyForQuery. All further processing of the query string is aborted by ErrorResponse (even if more queries remained in it). Note that this might occur partway through the sequence of messages generated by an individual query.

Extended query protocol docs says:

>When an error is detected while processing any extended-query message, the backend issues ErrorResponse, then reads and discards messages until a Sync is reached, then issues ReadyForQuery and returns to normal message processing. (But note that no skipping occurs if an error is detected while processing Sync — this ensures that there is one and only one ReadyForQuery sent for each Sync.)


In our case, our errors don't do recovery (they are all fatal), but we still need to send the ReadyForQuery.

If we don't send this ReadyForQuery, clients will get a bad-recv (tcp closed) while waiting for it. This happened with quite a few errors in Supavisor.

File: lib/supavisor/errors/max_connections_error.ex
Comment (v0idpwn) [reply lib/supavisor/errors/max_connections_error.ex:1]: I agree this one is awkward, I'll take a look on it. 

Previously we had two separate errors for full session pool and full txn pool. When I added full proxy, we got 3. This makes the caller code complex, so I'd prefer to have a single one. 

But maybe we should have 2 (max conn and max session conn) to reflect the different errors from a user perspective and having an abstraction to build them instead of making the error overly complex.
